rust/extractor/src/crate_graph.rs

```diff
@@ -0,0 +1,655 @@
+    let mut crate_id_map = HashMap::<CrateId, (PathBuf, u64)>::new();
+    for krate_id in crate_graph.crates_in_topological_order() {
+        let krate = &crate_graph[krate_id];
+        let root_module_file: &VfsPath = vfs.file_path(krate.root_file_id);
+        if let Some(root_module_file) = root_module_file
+            .as_path()
+            .map(|p| std::fs::canonicalize(p).unwrap_or(p.into()))
+        {
+            let mut hasher = std::hash::DefaultHasher::new();
+            krate
+                .cfg_options
+                .as_ref()
+                .into_iter()
+                .sorted_by(cmp_flag)
+                .for_each(|x| format!("{x}").hash(&mut hasher));
+
+            krate
+                .dependencies
+                .iter()
+                .flat_map(|d| crate_id_map.get(&d.crate_id))
+                .sorted()
+                .for_each(|x| x.hash(&mut hasher));
+            let hash = hasher.finish();
+            crate_id_map.insert(krate_id, (root_module_file, hash));
+        }
+    }
+    for krate_id in crate_graph.iter() {
+        if let Some((root_module_file, hash)) = crate_id_map.get(&krate_id) {
+            let path = root_module_file.join(format!("{hash:0>16x}"));
+            let mut trap = trap_provider.create("crates", path.as_path());
+            if trap.path.exists() {
+                continue;
+            }
+            let krate = &crate_graph[krate_id];
+            let element = generated::Crate {
+                id: trap::TrapId::Key(format!("crate:{}:{hash}", root_module_file.display())),
+                name: krate
+                    .display_name
+                    .as_ref()
+                    .map(|x| x.canonical_name().to_string()),
+                version: krate.version.to_owned(),
+                cfg_options: krate
+                    .cfg_options
+                    .as_ref()
+                    .into_iter()
+                    .map(|x| format!("{x}"))
+                    .collect(),
+                dependencies: krate
+                    .dependencies
+                    .iter()
+                    .flat_map(|x| crate_id_map.get(&x.crate_id))
+                    .map(|(module, hash)| {
+                        trap.label(format!("crate:{}:{hash}", module.display()).into())
+                    })
+                    .collect(),
+            };
+            let parent = trap.emit(element);
+
+            go(
+                &crate_graph,
+                db,
+                db.crate_def_map(krate_id).as_ref(),
+                parent.into(),
+                "crate",
+                DefMap::ROOT,
+                &mut trap,
+            );
+            trap.commit().unwrap_or_else(|err| {
+                log::error!(
+                    "Failed to write trap file for crate: {}: {}",
+                    root_module_file.display(),
+                    err.to_string()
+                )
+            });
+        }
+        fn go(
```

Comment:
I think we need a little more structure here, and extract parts of this function to smaller pieces. Right now it's very hard to follow what is happening here, this whole `extract_crate_graph` function body is just too big